### ISSUE TITLE
[ConvectionDiffusion] Constitutive law for heat problems?

### ISSUE THREAD

**Issue by takeshimg92** [KratosMultiphysics/Kratos#12795]
**Description**
I am adapting code using TableAccessors for variable material properties. In the StructuralMechanics application I had no problem, it amounts to rewriting `CalculatePK2Stress` and `CalculateElasticMatrix` inside the cpp files for the specific constitutive laws.

I want to do something similar for ConvectionDiffusion, e.g. if I want to specify thermal conductivity based on another variable. However, there is no equivalent of a constitutive law here. I am a bit lost on what I could do to add the TableAccessors.

**Scope**
ConvectionDiffusion application

**Comment by matekelemen** [comment]
Looks like the convdiff app does things in an ... unorthodox manner.
1) first of all, the constitutive laws are hard-coded into elements
2) it uses its [own class](https://github.com/KratosMultiphysics/Kratos/blob/3ecd8ee8c9fe789f5cd0acb5116639bf23ffcc32/kratos/includes/convection_diffusion_settings.h) for storing variables and material parameters. On the one hand, it's quite nice because it lets you tweak the diffusion variables to your specific case. On the other hand, it's not standard in Kratos and rather difficult to read.
3) **material parameters are stored as solution step variables**. I hope there's no good reason for this, otherwise your plan of adapting them to `Accessor`s will likely run into problems.

##

If I understand you correctly, you just want to fetch material parameters from `Accessor`s but don't want to change the constitutive law on the fly. If that's the case, it should be relatively straightforward: go through the elements you want changed and adapt each line where params are fetched.

For example, in `applications/ConvectionDiffusionApplication/custom_elements/laplacian_element.cpp`:

https://github.com/KratosMultiphysics/Kratos/blob/3ecd8ee8c9fe789f5cd0acb5116639bf23ffcc32/applications/ConvectionDiffusionApplication/custom_elements/laplacian_element.cpp#L103-L104

**Comment by loumalouomega** [comment]
> Looks like the convdiff app does things in an ... unorthodox manner.
> 
>     1. first of all, the constitutive laws are hard-coded into elements
> 
>     2. it uses its [own class](https://github.com/KratosMultiphysics/Kratos/blob/3ecd8ee8c9fe789f5cd0acb5116639bf23ffcc32/kratos/includes/convection_diffusion_settings.h) for storing variables and material parameters. On the one hand, it's quite nice because it lets you tweak the diffusion variables to your specific case. On the other hand, it's not standard in Kratos and rather difficult to read.
> 
>     3. **material parameters are stored as solution step variables**. I hope there's no good reason for this, otherwise your plan of adapting them to `Accessor`s will likely run into problems.
> 
> 
> If I understand you correctly, you just want to fetch material parameters from `Accessor`s but don't want to change the constitutive law on the fly. If that's the case, it should be relatively straightforward: go through the elements you want changed and adapt each line where params are fetched.
> 
> For example, in `applications/ConvectionDiffusionApplication/custom_elements/laplacian_element.cpp`:
> 
> [Kratos/applications/ConvectionDiffusionApplication/custom_elements/laplacian_element.cpp](https://github.com/KratosMultiphysics/Kratos/blob/3ecd8ee8c9fe789f5cd0acb5116639bf23ffcc32/applications/ConvectionDiffusionApplication/custom_elements/laplacian_element.cpp#L103-L104)
> 
> Lines 103 to 104 in [3ecd8ee](/KratosMultiphysics/Kratos/commit/3ecd8ee8c9fe789f5cd0acb5116639bf23ffcc32)
>  heat_flux_local[node_element] = r_geometry[node_element].FastGetSolutionStepValue(r_volume_source_var); 
>  nodal_conductivity[node_element] = r_geometry[node_element].FastGetSolutionStepValue(r_diffusivity_var);

Welcome to Kratos pre git...pre many things

**Comment by AlejandroCornejo** [comment]
@matekelemen is right, in this case, since no CL is defined for those elements, you need to modify the material properties directly at elements level but using the same procedures.

**Comment by takeshimg92** [comment]
Thank you all for the responses. For the time being, it seems just looping over nodes and SetSolutionStepValue'ing the desired variables works. Before closing this issue, I wanted to ask if there are any short-term plans of refactoring this application. Seems there are various patches to make it work.

**Comment by matekelemen** [comment]
I don't know of such plans unfortunately, at least not from the german teams (TU Munich / TU Braunschweig).

You're more than welcome to propose some changes though. The app seems relatively simple so I can at least promise to help you with basic stuff and review PRs if you open any.

**Comment by RiccardoRossi** [comment]
@rubenzorrilla you may want to comment here

**Comment by rubenzorrilla** [comment]
> [@rubenzorrilla](https://github.com/rubenzorrilla) you may want to comment here

Yes. I need to find the time to answer properly. Please leave it open for a while.

**Comment by rubenzorrilla** [comment]
It might seem unorthodox, specially if you take the `StructuralMechanicsApplication` as reference application in the comparison. However, everything has a reasoning behind
 
> 1. first of all, the constitutive laws are hard-coded into elements

In this regard, one should note that in here the unique "constitutive law" would be Fourier's law. Though one might consider non-linear conductivity, we've  never required this. Hence, and for the sake of simplicity and efficiency, the elements in `ConvectionDiffusionApplication` hardcode Fourier's law. Note that this is also done in other Kratos places.

Having said this, if required we can perfectly refactor the main elements (note that there are a few deprecated ones) to use constitutive laws. However, I'd only do this provided that there is a good reasoning behind (i.e., current implementation becomes in a usability limitation).

> 2. it uses its [own class](https://github.com/KratosMultiphysics/Kratos/blob/3ecd8ee8c9fe789f5cd0acb5116639bf23ffcc32/kratos/includes/convection_diffusion_settings.h) for storing variables and material parameters. On the one hand, it's quite nice because it lets you tweak the diffusion variables to your specific case. On the other hand, it's not standard in Kratos and rather difficult to read.

This you explained perfectly. The idea of such `ConvectionDiffusionSettings` is to use the very same implementation (elements) to solve any physical problem that boils down to a convection-diffusion equation (i.e., we can use the same implementation to both convect a pollutant and to solve a thermal problem).
 
> 3. **material parameters are stored as solution step variables**. I hope there's no good reason for this, otherwise your plan of adapting them to `Accessor`s will likely run into problems.> 

Unfortunately, there is a good reason for it. The idea is to enable solving problems with material properties that are non-constant in space. A different discussion is if we should use the non-historical database for this. In this regard, note that the selection is not easy. For instance, one might thing that `DENSITY` can be non-historical, but this would limit the coupling when using a weakly compressible fluid solver. Also note that the access to the historical database is faster (of course at the price of more memory allocation).

Long story short, this is a delicate discussion, not because the technicalities but because the resulting applicability. That being said, we can speak about which variables should remain nodal (historical/non-historical) and which ones should be property-based, and thus accessor-compatible.

**Comment by takeshimg92** [comment]
Many thanks @rubenzorrilla for the detailed answer. I have all information I need. For now, it is actually more convenient to have access to all properties as solution step variables, since I can modify them on the fly; not sure at this point if this will become a usability issue in the future. 

I will be closing the issue, thank you all for your help!
